src/EFCore.PG/Query/Internal/NpgsqlSqlTranslatingExpressionVisitor.cs

```diff
@@ -353,6 +353,108 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
+                || Visit(methodCallExpression.Arguments[0]) is not SqlExpression sqlIndexes)
+            {
+                return QueryCompilationContext.NotTranslatedExpression;
+            }
+
+            // For parameters and columns, construct the subquery: (SELECT array_agg(u.x + 1) FROM unnest(indexes) AS u(x))
+            // For constants and inline arrays, delegate to translator for optimization
+            if (sqlIndexes is SqlParameterExpression or ColumnExpression)
+            {
+                // Apply type mapping to the indexes array
+                var intArrayTypeMapping = _typeMappingSource.FindMapping(typeof(int[]))!;
+                sqlIndexes = _sqlExpressionFactory.ApplyTypeMapping(sqlIndexes, intArrayTypeMapping);
+
+                // Generate table alias and create unnest table
+                var tableAlias = ((RelationalQueryCompilationContext)_queryCompilationContext).SqlAliasManager.GenerateTableAlias("u");
+                var unnestTable = new PgUnnestExpression(tableAlias, sqlIndexes, "x", withOrdinality: false);
+
+                // Create column reference for unnested value
+                var intTypeMapping = _typeMappingSource.FindMapping(typeof(int))!;
+                var xColumn = new ColumnExpression("x", tableAlias, typeof(int), intTypeMapping, nullable: false);
+
+                // Create increment expression: x + 1
+                var xPlusOne = _sqlExpressionFactory.Add(xColumn, _sqlExpressionFactory.Constant(1, intTypeMapping));
+
+                // Create array_agg(x + 1) function
+                var arrayAggFunction = _sqlExpressionFactory.Function(
+                    "array_agg",
+                    [xPlusOne],
+                    nullable: true,
+                    argumentsPropagateNullability: new[] { true },
+                    typeof(int[]),
+                    intArrayTypeMapping);
+
+                // Construct SelectExpression
+#pragma warning disable EF1001 // SelectExpression constructors are pubternal
+                var selectExpression = new SelectExpression(
+                    [unnestTable],
+                    arrayAggFunction,
+                    [],
+                    ((RelationalQueryCompilationContext)_queryCompilationContext).SqlAliasManager);
+#pragma warning restore EF1001
+
+                // Finalize and wrap in ScalarSubqueryExpression
+                selectExpression.ApplyProjection();
+                var convertedIndexes = new ScalarSubqueryExpression(selectExpression);
+
+                // Delegate to translator to build the cube_subset function call
+                // by calling the base implementation which will invoke translators
+                return base.VisitMethodCall(
+                    methodCallExpression.Update(
+                        methodCallExpression.Object,
+                        [convertedIndexes]));
+            }
+
+            // For other cases (constants, inline arrays), let the translator handle it
```

Comment:
I'd honestly just handle all translations here (including constants/inline arrays) rather than split across here (for complex) and the translator (otherwise); but if you feel it's better the way it is I'm also OK with it.

I'd also extract this out to a private TranslateCubeToSubset method simply to get the complicated code out of here (this is definitely more complicated than the typical translation). But again, no strong feelings and I'm fine as-is as well.